yt_dlp/extractor/bilibili.py

```diff
@@ -774,3 +775,139 @@ def _real_extract(self, url):
         return self.url_result(
             'http://www.bilibili.tv/video/av%s/' % video_id,
             ie=BiliBiliIE.ie_key(), video_id=video_id)
+
+
+class BiliIntlBaseIE(InfoExtractor):
+    _API_URL = 'https://api.bili{}/intl/gateway'
+
+    def _call_api(self, endpoint, id):
+        return self._download_json(self._API_URL + endpoint, id)['data']
+
+    def _extract_formats_and_subtitles(self, ep_id):
+        video_json = self._call_api(f'/web/playurl?ep_id={ep_id}&platform=web', ep_id)
+        if not video_json:
+            self.raise_login_required(method='cookies')
+        video_json = video_json['playurl']
+        formats = []
+        for vid in video_json.get('video', []):
+            video_res = vid['video_resource']
+            video_info = vid['stream_info']
+            if video_res.get('url'):
+                formats.append({
+                    'url': video_res['url'],
+                    'ext': 'mp4',
+                    'format_note': video_info.get('desc_words'),
+                    'width': video_res.get('width'),
+                    'height': video_res.get('height'),
+                    'vbr': video_res.get('bandwidth'),
+                    'acodec': 'none',
+                    'vcodec': video_res.get('codecs'),
+                    'filesize': video_res.get('size'),
+                })
+        for aud in video_json.get('audio_resource', []):
+            if aud.get('url'):
+                formats.append({
+                    'url': aud['url'],
+                    'ext': 'mp4',
+                    'abr': aud.get('bandwidth'),
+                    'acodec': aud.get('codecs'),
+                    'vcodec': 'none',
+                    'filesize': aud.get('size'),
+                })
+        sub_json = self._call_api(f'/m/subtitle?ep_id={ep_id}&platform=web', ep_id)
+        subtitles = {}
+        for sub in sub_json.get('subtitles', []):
+            sub_url = sub.get('url')
+            if not sub_url:
+                continue
+            subtitles.setdefault(sub.get('key', 'en'), []).append({
+                'url': sub_url,
+            })
```

Comment:
formats and subtitle code seems completely seperate. plus, subtitles need an additional request. It is best to delegate this to `self._get_subtitles` and then call `self.extract_subtitles` in `_real_extract`

Reply:
why `_real_extract` and not `_extract_ep_info`?

Reply:
yeah, it should be in `_extract_ep_info`, my bad

Reply:
The new commit was made assuming this.